In [ ]:
library(tidyverse)
library(repr)
library(tidymodels)
library(dplyr)

In [92]:
github_players_url <- "https://raw.githubusercontent.com/1v9rstudioer/toy_ds_project/refs/heads/main/players.csv"
github_sessions_url <- "https://raw.githubusercontent.com/1v9rstudioer/toy_ds_project/refs/heads/main/sessions.csv"
players_data <- read_csv(github_players_url)
sessions_data <-read_csv(github_sessions_url)

players_data
sessions_data

Rows: 196 Columns: 7
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr (4): experience, hashedEmail, name, gender
dbl (2): played_hours, Age
lgl (1): subscribe

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
Rows: 1535 Columns: 5
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr (3): hashedEmail, start_time, end_time
dbl (2): original_start_time, original_end_time

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.


experience subscribe
1   Pro         TRUE    
2   Veteran     TRUE    
3   Veteran    FALSE    
4   Amateur     TRUE    
5   Regular     TRUE    
6   Amateur     TRUE    
7   Regular     TRUE    
8   Amateur    FALSE    
9   Amateur     TRUE    
10  Veteran     TRUE    
11  Veteran     TRUE    
12  Pro         TRUE    
13  Beginner    TRUE    
14  Amateur     TRUE    
15  Veteran     TRUE    
16  Amateur    FALSE    
17  Beginner    TRUE    
18  Amateur     TRUE    
19  Amateur     TRUE    
20  Regular     TRUE    
21  Regular     TRUE    
22  Amateur    FALSE    
23  Beginner    TRUE    
24  Beginner    TRUE    
25  Amateur     TRUE    
26  Regular     TRUE    
27  Regular     TRUE    
28  Veteran    FALSE    
29  Amateur     TRUE    
30  Veteran    FALSE    
⋮   ⋮          ⋮        
167 Amateur    FALSE    
168 Beginner   FALSE    
169 Veteran     TRUE    
170 Veteran    FALSE    
171 Pro        FALSE    
172 Beginner   FALSE    
173 Veteran     TRUE    
174 Regular    FALSE    
175 Veteran     TRUE    
176 Amateur     TRUE    
177 Pro        FALSE    
178 Veteran    FALSE    
179 Amateur     TRUE    
180 Veteran     TRUE    
181 Amateur    FALSE    
182 Amateur     TRUE    
183 Pro        FALSE    
184 Amateur     TRUE    
185 Pro        FALSE    
186 Regular    FALSE    
187 Veteran     TRUE    
188 Amateur     TRUE    
189 Beginner    TRUE    
190 Amateur    FALSE    
191 Amateur     TRUE    
192 Amateur     TRUE    
193 Veteran    FALSE    
194 Amateur    FALSE    
195 Amateur    FALSE    
196 Pro         TRUE    
    hashedEmail                                                     
1   f6daba428a5e19a3d47574858c13550499be23603422e6a0ee9728f8b53e192d
2   f3c813577c458ba0dfef80996f8f32c93b6e8af1fa939732842f2312358a88e9
3   b674dd7ee0d24096d1c019615ce4d12b20fcbff12d79d3c5a9d2118eb7ccbb28
4   23fe711e0e3b77f1da7aa221ab1192afe21648d47d2b4fa7a5a659ff443a0eb5
5   7dc01f10bf20671ecfccdac23812b1b415acd42c2147cb0af4d48fcce2420f3e
6   f58aad5996a435f16b0284a3b267f973f9af99e7a89bee0430055a44fa92f977
7   8e594b8953193b26f498db95a508b03c6fe1c24bb5251d392c18a0da9a722807
8   1d2371d8a35c8831034b25bda8764539ab7db0f63938696917c447128a2540dd
9   8b71f4d66a38389b7528bb38ba6eb71157733df7d1740371852a797ae97d82d1
10  bbe2d83de678f519c4b3daa7265e683b4fe2d814077f9094afd11d8f217039ec
11  7a4686586d290c67179275c7c3dfb4ea02f4d317d9ee0e2cee98baa27877a875
12  4caa42e1b20511552434978171dcf7283fb6eb857eb871327e39bfc328002309
13  4bed6f49b672974f706aa7f020b4f29020326bf594520593557cd381a51559b8
14  3caa832978e0596779f4ee7c686c4592fb6de893925025d25bf26be9a8cd03f5
15  6f9acf8ea9956fe817895c78d10e1e25c11aba335a451e5b0322b073ee5008c4
16  2313a06afe47eacc28ff55adf6f072e7d12b0d12d7cbae29656f9029ec72bb53
17  9db37a1d381697f6c83e3b97b276e38bb9d9c811ba4596edb75d4dbfea6415b0
18  a175d4741dc84e6baf77901f6e8e0a06f54809a34e6b5211159bced346f7fb3e
19  ab1f44f93c3b828f55458971db393052d9711df3e0e7ff69540bfebfcec555ff
20  20ae3f8b5323674fe6e537bc91928841cfc2a00c1b36522fd25370bb6fa89d36
21  bc704ff2bc676dbf48ee41b9e11481c1387bf758ad318f2428f336e3fecc6660
22  200db0417ee70863c7e2a0f214c737476fb86de5869d0b907a077d726357e687
23  f9ac013b2f0bc2bd4928a6a0fc8a0aae8b5c4f2670cf9e18f547b7e72034c965
24  89c8e07d6a2aa102d9cd05086d51b2e41195c9c807dd976a06f4262ff36d0cca
25  119f01b9877fc5ea0073d05602a353b91c4b48e4cf02f42bb8d661b46a34b760
26  5baba1651a0b92788bc0d6dcdf00be64af1cf9f0015bbeed204ddabd1d872e83
27  50cec73a632e4be7b155c3df5165de28b2c5baf092abcb7f509f5645776a56c2
28  f8acd26a5e452b682b9f8b20108caef1ee2e745efe08e95806980bd2388a3fea
29  4b01bce3f141289709e8278b02ba5d2aaa7105d7ccb9c7deb37670a80e332774
30  951e54f7376e2b2f0915e9e3646c701af4a2fe839385b13f9e0084b08f8070bd
⋮   ⋮                                                               
167 7096b30972a32029244fc6171f21bc5996b51b8d75df7a94f003dead428988f9
168 42eafe96ed5c1684e3b5cc614d1b01a117173d3ec6898abcf1177d9822bd39f7
169 9ba67e127e4e70d909f6237644bb4674b17bc6ae5c9e735f88cfc50e22b499c6
170 4bfad3613c71ace05644bf210195d9fb0d3

hashedEmail                                                     
1    bfce39c89d6549f2bb94d8064d3ce69dc3d7e72b38f431d8aa0c4bf95ccee6bf
2    36d9cbb4c6bc0c1a6911436d2da0d09ec625e43e6552f575d4acc9cf487c4686
3    f8f5477f5a2e53616ae37421b1c660b971192bd8ff77e3398304c7ae42581fdc
4    bfce39c89d6549f2bb94d8064d3ce69dc3d7e72b38f431d8aa0c4bf95ccee6bf
5    36d9cbb4c6bc0c1a6911436d2da0d09ec625e43e6552f575d4acc9cf487c4686
6    bfce39c89d6549f2bb94d8064d3ce69dc3d7e72b38f431d8aa0c4bf95ccee6bf
7    fd6563a4e0f6f4273580e5fedbd8dda64990447aea5a33cbb5e894a3867ca44d
8    ad6390295640af1ed0e45ffc58a53b2d9074b0eea694b16210addd44d7c81f83
9    96e190b0bf3923cd8d349eee467c09d1130af143335779251492eb4c2c058a5f
10   36d9cbb4c6bc0c1a6911436d2da0d09ec625e43e6552f575d4acc9cf487c4686
11   fd6563a4e0f6f4273580e5fedbd8dda64990447aea5a33cbb5e894a3867ca44d
12   f8f5477f5a2e53616ae37421b1c660b971192bd8ff77e3398304c7ae42581fdc
13   f2826fb8dbce4d450348f99cb27ade184b713998d9679780442efaaf218038f2
14   b622593d2ef8b337dc554acb307d04a88114f2bf453b18fb5d2c80052aeb2319
15   24d4892c2c3ce11d3e54c3bf31ee218901cac7a2f564911d67424c8fdf0874fe
16   fd6563a4e0f6f4273580e5fedbd8dda64990447aea5a33cbb5e894a3867ca44d
17   88247d9a46fc214a12485dcbcbb03a8ddebfe8c1ec5fe209cd660147625c8e62
18   f6daba428a5e19a3d47574858c13550499be23603422e6a0ee9728f8b53e192d
19   36d9cbb4c6bc0c1a6911436d2da0d09ec625e43e6552f575d4acc9cf487c4686
20   fd6563a4e0f6f4273580e5fedbd8dda64990447aea5a33cbb5e894a3867ca44d
21   bfce39c89d6549f2bb94d8064d3ce69dc3d7e72b38f431d8aa0c4bf95ccee6bf
22   36d9cbb4c6bc0c1a6911436d2da0d09ec625e43e6552f575d4acc9cf487c4686
23   fd6563a4e0f6f4273580e5fedbd8dda64990447aea5a33cbb5e894a3867ca44d
24   ad6390295640af1ed0e45ffc58a53b2d9074b0eea694b16210addd44d7c81f83
25   ad6390295640af1ed0e45ffc58a53b2d9074b0eea694b16210addd44d7c81f83
26   fd6563a4e0f6f4273580e5fedbd8dda64990447aea5a33cbb5e894a3867ca44d
27   36d9cbb4c6bc0c1a6911436d2da0d09ec625e43e6552f575d4acc9cf487c4686
28   bfce39c89d6549f2bb94d8064d3ce69dc3d7e72b38f431d8aa0c4bf95ccee6bf
29   b622593d2ef8b337dc554acb307d04a88114f2bf453b18fb5d2c80052aeb2319
30   fd6563a4e0f6f4273580e5fedbd8dda64990447aea5a33cbb5e894a3867ca44d
⋮    ⋮                                                               
1506 ad6390295640af1ed0e45ffc58a53b2d9074b0eea694b16210addd44d7c81f83
1507 bfce39c89d6549f2bb94d8064d3ce69dc3d7e72b38f431d8aa0c4bf95ccee6bf
1508 bfce39c89d6549f2bb94d8064d3ce69dc3d7e72b38f431d8aa0c4bf95ccee6bf
1509 b3510c708bd50bf9f75e6e02bb6fe14edb705e0ea671ee769a45c613750e8b58
1510 bfce39c89d6549f2bb94d8064d3ce69dc3d7e72b38f431d8aa0c4bf95ccee6bf
1511 e74c60a92c0100e7240be56d66969db85856152b048c639331a3595f901461ae
1512 b622593d2ef8b337dc554acb307d04a88114f2bf453b18fb5d2c80052aeb2319
1513 f8f5477f5a2e53616ae37421b1c660b971192bd8ff77e3398304c7ae42581fdc
1514 e74c60a92c0100e7240be56d66969db85856152b048c639331a3595f901461ae
1515 fd6563a4e0f6f4273580e5fedbd8dda64990447aea5a33cbb5e894a3867ca44d
1516 f8f5477f5a2e53616ae37421b1c660b971192bd8ff77e3398304c7ae42581fdc
1517 a1e0438d568b9c05a93cadaf99969e4d44794d19dea31edd723cbc916f878c33
1518 fd6563a4e0f6f4273580e5fedbd8dda64990447aea5a33cbb5e894a3867ca44d
1519 36d9cbb4c6bc0c1a6911436d2da0d09ec625e43e6552f575d4acc9cf487c4686
1520 55d24216db39c27e1f17cc43d3127cbf8ed76ada6d098202b53ded319855e2c1
1521 f8f5477f5a2e53616ae37421b1c660b971192bd8ff77e3398304c7ae42581fdc
1522 b622593d2ef8b337dc554acb307d04a88114f2bf453b18fb5d2c80052aeb2319
1523 b622593d2ef8b337dc554acb307d04a88114f2bf453b18fb5d2c80052aeb2319
1524 f8f5477f5a2e53616ae37421b1c660b971192bd8ff77e3398304c7ae42581fdc
1525 fd6563a4e0f6f4273580e5fedbd8dda64990447aea5a33cbb5e894a3867ca44d
1526 18936844e06b6c7871dce06384e2d142dd86756941641ef39cf40a9967ea14e3
1527 f8f5477f5a2e53616ae37421b1c660b971192bd8ff77e3398304c7ae42581fdc
1528 18936844e06b6c7871dce06384e2d142dd86756941641ef39cf40a9967ea14e3
1529 f8f5477f5a2e53616ae37421b1c660b971192bd8ff77e3398304c7ae42581fdc
1530 b622593d2ef8b337dc554acb307d04a88114f2bf453b18fb5d2c80052aeb2319
1531 36d9cbb4c6

In [ ]:
#tidying the data
players <- players_data |>
    mutate(
    experience = as.factor(experience),
    subscribe = as.logical(subscribe),
    gender = as.factor(gender)
        )

#we dont need any identifying columns for now, so we'll remove them
players <- players |> select(-hashedEmail, -name)

summary(players)

sessions <- sessions_data |>
  mutate(
    start_time = dmy_hm(start_time),
    end_time = dmy_hm(end_time),
    original_start_time = as_datetime(original_start_time / 1000),
    original_end_time = as_datetime(original_end_time / 1000)
  )

#missing values check
colSums(is.na(sessions))

#calculate session duration in minutes
sessions <- sessions_data |>
  mutate(session_duration = difftime(original_end_time, original_start_time, units = "mins"))

summary(sessions)


1. Data description:

There are **196 observations** (rows) total in the **players.csv** dataset, including quantitative data on playtime alongside personal identity information like age and gender.

The "players.csv" dataset includes 7 columns and variables:
- experience: Pro, Veteran, Regular, Beginner, or Amateur. *Majority of players are Ameteur at 63*
- subscribe: TRUE/FALSE if a player subscribed to game-related newsletter. *52 arent, 144 are.*
- hashedEmail: encrypted personal email of player (for identification)
- played_hours: number of hours played. *Mean of total hours played is 5.846 hours*
- name: name of player
- gender: string label of gender. Out of all the players, *124 are male, 37 female, 15 NB, 6 TS, 1 Other, 2 Agender*, and notably **11 people did not state thier gender**
- Age: age of player, mean is *20.52* notably, **2 people neglected to state thier age**

A potential issue I see arising in the data is with missing data, particularly 2 observations in age and 11 in gender, which may make it difficult to answer the 1st reserch question regarding player characteristics and subscribing to the newsletter. Additionally, playtime has a huge outlier of 223hrs, which is extremely far off from the mean suggesting the dataset may include statistical landmines. Another issue could be in the nature of how the data was recorded. Were players observed by a panel to determine thier skill level or did they self-assess/report thier skill? How accurate can such a measure be? This may introduce bias into the dataset.

There are a total of **1535 observations** in the **sessions.csv** dataset, which includes a large amount of quantitiave data about timestamps of specific user sign-on sessions to the PLAIcraft server and a encrypted email for identification

The columns/variables are:
-hashedEmail: identification of user signing onto server
-start_time: Gaming session start time, converted from direct UNIX timestamp in miliseconds to DY/MN/YEAR + Time (24hrs, string)
-end_time: Gaming session end time, converted from direct UNIX timestamp to DY/MN/YEAR + Time (24hrs, string)
-original_start_time: original UNIX timestamp for session start
-original_end_time: original UNIX timestamp for session end

125 unique players were recorded, and the most frequent player had 310 sessions of play. No statistics on session time could be computed as of now, as the converted time is a string, and the UNIX time columns are in milliseconds. Notably, there are **two session end timestamps** which are missing from the dataset. 
    


2. Addressing a broad question with specific questions. What trends/associations can we examine?

My specifc question will be inregards to the broader question #3: 

"Can session start times and player activity patterns predict peak concurrent player counts on the Minecraft research server?"

Im sure people who play minecraft together want to minimize the time playing alone, and maximize the time playing together. This is definitely what I would like to do, but I know myself and many others are annoyed when we are told to hop on at a specific time, just to have the majority of our friends log off 30 minutes later. Just because someone says they will start at a certain time, doesnt mean you will play with them around that time. So we will attempt to identify whether the session start time can really indicate peak player activity, for the future of time management and optimizing procrastination!

The way we will execute this is by first extracting timestamps into a standard Date/time format and extract key time based features such as hour of day, day of week, and month. We can aggregate concurrent sessions for each hourly time window to form a  dataset where each row represents an hour and the response variable is the number of concurrent players. Later, we can use a multiple linear regresion model to predict concurrent player activity based on extracted features like time of day or day of week, alongside session duration. Then we will evaluate the model using RMSE to measure how well our predictions match actual player counts

3. Data Analysis and Visualization

We have already done a little rudimentary statistical summaries on the players.csv dataset, but we will reload them cuz why not. Additionally, we will create a few exploratory visualizations that may help us structure and focus or research question some more.

| Experience Level | Subscribe (Mode) | Played Hours (Min, Q1, Median, Mean, Q3, Max) | Gender Distribution                         | Age (Min, Q1, Median, Mean, Q3, Max, NAs) |
|-----------------|-----------------|-----------------------------------------------|--------------------------------------------|--------------------------------------------|
| Amateur: 63     | FALSE: 52        | Min: 0.000                                  | Agender: 2                                 | Min: 8.00                                 |
| Beginner: 35    | TRUE: 144        | Q1: 0.000                                   | Female: 37                                 | Q1: 17.00                                 |
| Pro: 14        |                 | Median: 0.100                               | Male: 124                                  | Median: 19.00                             |
| Regular: 36     |                 | Mean: 5.846                                 | Non-binary: 15                             | Mean: 20.52                               |
| Veteran: 48     |                 | Q3: 0.600                                   | Other: 1                                   | Q3: 22.00                                 |
|                 |                 | Max: 223.100                                | Prefer not to say: 11                      | Max: 50.00                                |
|                 |                 |        


We had to do some additional wranging to our data, to make our visualizations possible (because of the time-nature of the predictors we are measuring). Our exploratory visualizations begin by converting  the original_start_time and original_end_time to proper date-time objects: hour_of_day and day_of_week and month. The original_start_time and original_end_time columns store timestamps in UNIX milliseconds (e.g., 1638316800000), so we had to convert them by dividing by 1000 and using as.POSIXct(). Calculating the length of each session was obtained by subtracting start_time from end_time. We then removed the rows with missing timestamps as mentioned earlier after checking for them, then created a histogram for the distribution of gaming sessions throughout the day (24hrs), a bar plots for the 7 days of the week, and a bar plot for the distribution of start times throughout the availible months. Additionally, we added a histogram to display how long gaming sessions typically last, helping us understand if players prefer short periods or longer periods of play (in progress, couldnt make it work in time for the deadline). The visualizations reveal clear patterns in player activity: peak hours (e.g., late afternoon/evening), higher activity on weekends, and potential seasonal trends (e.g., more play in summer). These insights directly support our research question by showing that time-based features (hour, day, month) are strong predictors of player activity.

In [ ]:
sessions_data <- sessions_data %>%
  mutate(
    start_time = as.POSIXct(original_start_time / 1000, origin = "1970-01-01", tz = "UTC"),
    end_time = as.POSIXct(original_end_time / 1000, origin = "1970-01-01", tz = "UTC")
  )

sessions_data <- sessions_data %>%
  mutate(
    hour_of_day = hour(start_time), # Extract hour of the day
    day_of_week = wday(start_time, label = TRUE, abbr = FALSE), # Extract day of the week
    month = month(start_time, label = TRUE, abbr = FALSE) # Extract month
  )

sessions_data <- sessions_data %>%
  mutate(session_duration = as.numeric(difftime(end_time, start_time, units = "mins")))

sum(is.na(sessions_data$original_start_time))
sum(is.na(sessions_data$original_end_time))

invalid_durations <- sessions_data %>%
  filter(end_time < start_time)
nrow(invalid_durations)

sessions_data_clean <- sessions_data %>%
  filter(!is.na(original_start_time)) %>%
  filter(!is.na(original_end_time)) %>%
  filter(end_time >= start_time)

ggplot(sessions_data_clean, aes(x = hour_of_day)) +
  geom_histogram(binwidth = 1, fill = "skyblue", color = "black") +
  labs(
    title = "Distribution of Session Start Times by Hour of the Day",
    x = "Hour of the Day",
    y = "Number of Sessions"
  ) +
  theme_minimal()

ggplot(sessions_data_clean, aes(x = day_of_week)) +
  geom_bar(fill = "lightgreen", color = "black") +
  labs(
    title = "Distribution of Session Start Times by Day of the Week",
    x = "Day of the Week",
    y = "Number of Sessions"
  ) +
  theme_minimal()

ggplot(sessions_data_clean, aes(x = month)) +
  geom_bar(fill = "orange", color = "black") +
  labs(
    title = "Distribution of Session Start Times by Month",
    x = "Month",
    y = "Number of Sessions"
  ) +
  theme_minimal()



4. How will we address our question of interest?

Linear regression is a suitable analysis method because the connection between the variables we are measuring are quantitative (time-based measures and concurrent player count) and Linear regression specializes in creating predictions of future trends based on past variable trends. For this model to accurately make predictions, we must assume there an approximately linear relationship between the predictors (time metrics) and response variable (concurrent player activity). A limitation of this assumption is that if play activity exhibits non-linear trends such as peak hours exponentially growing in player count, they will be missed by the model. First, we will have to convert a lot of our time data (which is now currently in microseconds) into specific objects, hours, days of week, and session duration all require wrangling the data into new datasets or updating the existing ones.To evaluate the effectivenes of the model in predicting player activity, we will use RMSE to asssess the accuracy on a 30% test set split from a 70% training set we will create the model on. Cross validation will be executed on the training set to ensure data stability and avoid overfitting, with 5 folds used as a validation set once during training. R^2 value can also provide a measure of variance explainable by the model, and with our variables we should expect something 